tiledb/api/src/query_arrow/arrow.rs

```diff
@@ -0,0 +1,520 @@
+
+use thiserror::Error;
+
+use crate::array::schema::CellValNum;
+use crate::datatype::Datatype;
+
+#[derive(Error, Debug, PartialEq, Eq)]
+pub enum Error {
+    #[error("Cell value size '{0}' is out of range.")]
+    CellValNumOutOfRange(u32),
+
+    #[error("Internal type error: Unhandled Arrow type: {0}")]
+    InternalTypeError(adt::DataType),
+
+    #[error("Invalid fixed sized length: {0}")]
+    InvalidFixedSize(i32),
+
+    #[error("Invalid Arrow type for conversion: '{0}'")]
+    InvalidTargetType(adt::DataType),
+
+    #[error("Failed to convert Arrow list element type: {0}")]
+    ListElementTypeConversionFailed(Box<Error>),
+
+    #[error(
+        "The TileDB datatype '{0}' does not have a default Arrow DataType."
+    )]
+    NoDefaultArrowType(Datatype),
+
+    #[error("Arrow type '{0} requires the TileDB field to be single valued.")]
+    RequiresSingleValued(adt::DataType),
+
+    #[error("Arrow type {0} requires the TileDB field be var sized.")]
+    RequiresVarSized(adt::DataType),
+
+    #[error("TileDB does not support timezones on timestamps")]
+    TimeZonesNotSupported,
+
+    #[error(
+        "TileDB type '{0}' and Arrow type '{1}' have different physical sizes"
+    )]
+    PhysicalSizeMismatch(Datatype, adt::DataType),
+
+    #[error("Unsupported Arrow DataType: {0}")]
+    UnsupportedArrowDataType(adt::DataType),
+
+    #[error("TileDB does not support lists with element type: '{0}'")]
+    UnsupportedListElementType(adt::DataType),
+
+    #[error("The Arrow DataType '{0}' is not supported.")]
+    ArrowTypeNotSupported(adt::DataType),
+    #[error("DataFusion does not support multi-value cells.")]
+    InvalidMultiCellValNum,
+    #[error("The TileDB Datatype '{0}' is not supported by DataFusion")]
+    UnsupportedTileDBDatatype(Datatype),
+    #[error("Variable-length datatypes as list type elements are not supported by TileDB")]
+    UnsupportedListVariableLengthElement,
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// ConversionMode dictates whether certain conversions are allowed
+pub enum ConversionMode {
+    /// Only allow conversions that are semantically equivalent
+    Strict,
+    /// Allow conversions as long as the physical type is maintained.
+    Relaxed,
+}
+
+pub struct ToArrowConverter {
+    mode: ConversionMode,
+}
+
+impl ToArrowConverter {
+    pub fn strict() -> Self {
+        Self {
+            mode: ConversionMode::Strict,
+        }
+    }
+
+    pub fn physical() -> Self {
+        Self {
+            mode: ConversionMode::Relaxed,
+        }
+    }
+
+    pub fn convert_datatype(
+        &self,
+        dtype: &Datatype,
+        cvn: &CellValNum,
+        nullable: bool,
+    ) -> Result<adt::DataType> {
+        if let Some(arrow_type) = self.default_arrow_type(dtype) {
+            self.convert_datatype_to(dtype, cvn, nullable, arrow_type)
+        } else {
+            Err(Error::NoDefaultArrowType(*dtype))
+        }
+    }
+
+    pub fn convert_datatype_to(
```

Comment:
This function looks to me like it is written only with the default datatype case in mind, so it's missing some things.
- the user might request a List type which isn't covered here
- the binary/utf-8 case looks too permissive in that it allows i64 attributes; but not permissive enough in that fixed cell val num also oughta work (and we would later have to handle the offsets)

Reply:
I reckon the default type selection comes in handy for fixed cell val num, since then FixedSizeBinary is clearly the right thing.  But if the user requests UTF-8 from your fixed size attribute (which is definitely a tables-realm scenario) then I'd say it should be treated as compatible and the query can just whip up some offsets easily.

The other direction, going from var cell val num into a FixedSizeBinary, isn't a scenario I expect we will encounter, so we can skip it.  But I think allocating space for offsets and writing them and checking that they are all the same length would be a fine thing to do.

Reply:
You've confused me on this one. For `FixedSizeBinary` we've got a direct data type that'd cover all cases (though I wonder what it'd do for CellValNum::single()). So I'd expect that to map directly 100% of the time.

For strings, as you point out, we can just generate our own offsets for read compatibility. You mentioned that writing fixed length binary is unlikely (though will be transparently supportable). Its the string side that worries me when folks attempt to write arbitrary strings into a fixed width cell. I'm fairly certain it would be supportable, just in a "lets kick that off into its own PR" after we're sure how we want to do it as I could see a number of approaches on where that conversion/enforcement happens.